### PR TITLE
Bless "ReferenceGraph" by putting it in astmodel

### PR DESCRIPTION
**Pull request by Porges**
groundwork!

### CODE REVIEW

File: hack/generator/pkg/astmodel/reference_graph.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+	references := make(map[TypeName]TypeNameSet)
+	for _, def := range types {
+		references[def.Name()] = def.References()
+	}
+
+	return NewReferenceGraph(roots, references)
+}
+
+// Connected returns the set of types that are reachable from the
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/reference_graph.go:47]: super duper minor: You break early here but then down below a comment extends further.

Net is I think this comment could be on one line.

File: hack/generator/pkg/astmodel/reference_graph.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	for node := range c.roots {
+		c.CollectTypes(node, connectedTypes)
+	}
+
+	return connectedTypes
+}
+
+// CollectTypes returns the set of types that are reachable from the given
+// 'node' (and places them in 'collected').
+func (c ReferenceGraph) CollectTypes(node TypeName, collected TypeNameSet) {
```
Comment (matthchr) [inline hack/generator/pkg/astmodel/reference_graph.go:61]: Does this need to be exported?

Comment (Porges) [reply hack/generator/pkg/astmodel/reference_graph.go:61]: no, not even for the next "step" (#202) :+1:

File: hack/generator/pkg/astmodel/reference_graph.go (hunk 1 above)
Comment (Porges) [reply hack/generator/pkg/astmodel/reference_graph.go]: done

File: hack/generator/pkg/astmodel/reference_graph.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+// NewReferenceGraphWithResourcesAsRoots produces a ReferenceGraph for the given set of
+// types, where the Resource types are the roots.
+func NewReferenceGraphWithResourcesAsRoots(types Types) ReferenceGraph {
+	roots := CollectResourceDefinitions(types)
+
+	references := make(map[TypeName]TypeNameSet)
+	for _, def := range types {
+		references[def.Name()] = def.References()
+	}
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/reference_graph.go:42]: My kingdom for LINQ. #again

**Pull request merged**